### PR TITLE
Address issue #323 and enable GHE Auth

### PR DESCRIPTION
**Pull request by tortilaman**
So far this is a basic fix for issue #323. Unauthorized users will not be able to get into the backend, and an alert window will popup informing them they need to speak with their site administrator to get access.

Once I figure out how to feed the error to AuthenticationPage.js I’ll make it look better.

Also, api_root wasn’t being fed to API.js, which was causing issues authenticating Github Enterprise users, since it was using the regular github api endpoint. I fed api_root in, which should enable Github Enterprise for everyone, and potentially other git providers?

### CODE REVIEW

**Comment by erquhart** [comment]
@tortilaman fyi I rebased your branch so that only the new commits are included in this PR.

File: .eslintrc [hunk 1]
```diff
@@ -11,5 +11,10 @@
   },
   "globals": {
     "CMS_ENV": true
+  },
+  "parserOptions": {
```
Comment (erquhart) [inline .eslintrc]: Remove this for now, can be brought up in a separate PR.

**Comment by tortilaman** [comment]
Yeah, I was gonna ask about rebasing. Figured that would be a good idea. I've actually now modified to provide better feedback so it uses the existing toast UI and redux-notifications. I feel comfortable merging at this point.

**Comment by tortilaman** [comment]
Apologies for the double commit there, was using github desktop instead of cli and sometimes it doesn't make it clear what it's doing.

**Comment by erquhart** [comment]
Fixed history again - if you do a hard reset to origin from your local repo, that should prevent further extraneous commits.

File: src/backends/github/implementation.js [hunk 2]
```diff
@@ -29,11 +30,20 @@ export default class GitHub {
...
+      this.api.collaborator(user.login).then((status) => {
+        if (status === 404 || status === 403) {
+          // Unauthorized user
+          throw new Error("Sorry, you don't have CMS access.");
+        } else if (status === 204) {
+          // Authorized user
+          user.token = state.token;
+          return user;
+        }
+        throw new Error('No response from Github, that\'s odd.');
```
Comment (erquhart) [inline src/backends/github/implementation.js]: Replace with "GitHub is not responding, please try again."

File: src/backends/github/implementation.js [hunk 3]
```diff
@@ -29,11 +30,20 @@ export default class GitHub {
...
-    return this.api.user().then((user) => {
-      user.token = state.token;
-      return user;
-    });
+    this.api = new API({ token: this.token, branch: this.branch, repo: this.repo, api_root: this.api_root });
+    return this.api.user().then(user =>
+      this.api.collaborator(user.login).then((status) => {
+        if (status === 404 || status === 403) {
+          // Unauthorized user
+          throw new Error("Sorry, you don't have CMS access.");
```
Comment (erquhart) [inline src/backends/github/implementation.js]: Replace with "Your GitHub user account does not have access to this repo."

File: src/backends/github/implementation.js (hunk 2 above)
Comment (tortilaman) [inline src/backends/github/implementation.js]: That probably makes more sense, haha. Leftover from my initial testing

File: src/backends/github/implementation.js (hunk 3 above)
Comment (tortilaman) [inline src/backends/github/implementation.js]: I was thinking that it might be beneficial in terms of users' mental model to abstract it a bit from GitHub since I'm sure there's a lot of end-users doing editing and writing that might not necessarily be so familiar with GitHub.

File: src/backends/github/implementation.js [hunk 4]
```diff
@@ -35,7 +35,11 @@ export default class GitHub {
       this.api.collaborator(user.login).then((status) => {
         if (status === 404 || status === 403) {
           // Unauthorized user
+<<<<<<< HEAD
```
Comment (Benaiah) [inline src/backends/github/implementation.js]: Looks like you inadvertently committed a diff here.

**Comment by tortilaman** [comment]
Everything should be cleaned up now for a PR and the diff should've been addressed.

File: src/backends/github/AuthenticationPage.js [hunk 5]
```diff
@@ -2,11 +2,13 @@ import React from 'react';
...
 import Authenticator from '../../lib/netlify-auth';
 import { Icon } from '../../components/UI';
+import { Notifs } from 'redux-notifications';
+import { Toast } from '../../components/UI/index';
 import styles from './AuthenticationPage.css';
 
 export default class AuthenticationPage extends React.Component {
   static propTypes = {
-    onLogin: React.PropTypes.func.isRequired,
+    onLogin: React.PropTypes.func.isRequired
```
Comment (Benaiah) [inline src/backends/github/AuthenticationPage.js]: This is an unrelated change, and goes against our current `eslint` config.

File: src/backends/github/API.js [hunk 6]
```diff
@@ -19,6 +19,10 @@ export default class API {
     return this.request("/user");
   }
 
+  collaborator(user) {
```
Comment (Benaiah) [inline src/backends/github/API.js]: This method should be renamed and modified as per my refactoring suggestions below.

File: src/actions/auth.js [hunk 7]
```diff
@@ -60,6 +63,11 @@ export function loginUser(credentials) {
         dispatch(authenticate(user));
       })
       .catch((error) => {
+        dispatch(notifSend({
```
Comment (Benaiah) [inline src/actions/auth.js]: :+1:, this is a long-awaited change.

File: .gitignore [hunk 8]
```diff
@@ -6,3 +6,6 @@ npm-debug.log
 .tern-project
 yarn-error.log
 .vscode/
+manifest.yml
+.imdone/
+config.local.yml
```
Comment (Benaiah) [inline .gitignore]: Don't know how I feel about these changes - on one hand, the .gitignore isn't super important; OTOH, I'd rather keep it small as it can cause confusion if it inadvertently ends up ignoring paths that we end up using, and if we add ignores for each contributor's individual tooling (rather than those that every contributor needs to ignore, such as build artifacts), I can see the `.gitignore` becoming too large pretty quickly. Additionally, these changes show up in history and diffs despite being unrelated to the actual purpose of the PR. @erquhart thoughts?

File: src/actions/auth.js [hunk 9]
```diff
@@ -1,4 +1,7 @@
 import { currentBackend } from '../backends/backend';
+import { actions as notifActions } from 'redux-notifications';
+
+const { notifSend } = notifActions;
```
Comment (Benaiah) [inline src/actions/auth.js]: Today I learned you can't do full destructuring in an `import` call - I was going to suggest `import { actions: { notifSend } } from 'redux-notifications';`, but that syntax is only supported for assignment, not imports).

File: src/backends/github/API.js [hunk 10]
```diff
@@ -64,7 +68,9 @@ export default class API {
     return fetch(url, { ...options, headers }).then((response) => {
       responseStatus = response.status;
       const contentType = response.headers.get("Content-Type");
-      if (contentType && contentType.match(/json/)) {
+      if (url.indexOf('/collaborators/') >= 0) {
+        return responseStatus;
+      } else if (contentType && contentType.match(/json/)) {
```
Comment (Benaiah) [inline src/backends/github/API.js]: While I understand the motivation behind this code, this is bound to cause bugs in the future by special-casing a particular URL in what is otherwise a generic request function. This should really be done in `collaborator`. One issue I see with the `response` method as it stands is that JSON responses reject the promise when an error status is present, but text responses do not.

@erquhart if you agree that that should be fixed, I can include it in my GH API refactor branch I'm preparing to PR (the branch name is `gh-api-refactors`, there's a little there already).

@tortilaman for now, you should be able to expect a JSON response from the GH API. Thus, any non-200 statuses will cause the promise to reject with the `APIError` in the `catch` block just below these lines. In addition,  My suggestions for refactoring this are as follows:

- Remove this logic from `request`.
- Rename the `collaborator` method to `isCollaborator`, and have that return a promise that either rejects or resolves to a boolean, representing whether the user is a collaborator or not.
- Catch the `404`, `403`, and `204` statuses in `isCollaborator`. If the status is one of those three, resolve the promise to the appropriate boolean. Otherwise, rethrow the error. (remember, JSON responses reject the error if the status is not 200-99, so you don't need any additional 
- Remove the now-unnecessary status code checks from `authenticate` and change them to handle the boolean that `isCollaborator` is returning. Add a `catch` block to handle unexpected errors, and move the `throw new Error('GitHub is not responding, please try again.');` to the `catch` block.

I'd be happy to elaborate on this further if necessary, and I can always refactor this myself if you'd rather move on, but either way I'd like this to be refactored before a merge.

File: src/backends/github/API.js [hunk 11]
```diff
@@ -236,7 +242,6 @@ export default class API {
   persistFiles(entry, mediaFiles, options) {
     const uploadPromises = [];
     const files = mediaFiles.concat(entry);
-
```
Comment (Benaiah) [inline src/backends/github/API.js]: This change is unrelated to the PR.

File: src/backends/github/API.js [hunk 12]
```diff
@@ -343,7 +348,7 @@ export default class API {
 
   deleteUnpublishedEntry(collection, slug) {
     const contentKey = slug;
-    let prNumber; 
+    let prNumber;
```
Comment (Benaiah) [inline src/backends/github/API.js]: This change, while good, is unrelated to the PR.

File: src/backends/github/implementation.js [hunk 13]
```diff
@@ -29,11 +30,20 @@ export default class GitHub {
...
+    return this.api.user().then(user =>
+      this.api.collaborator(user.login).then((status) => {
+        if (status === 404 || status === 403) {
+          // Unauthorized user
+          throw new Error("Your GitHub user account does not have access to this repo.");
+        } else if (status === 204) {
+          // Authorized user
+          user.token = state.token;
+          return user;
+        }
```
Comment (Benaiah) [inline src/backends/github/implementation.js]: The preceding status-checking logic should move to `collaborator`/`isCollaborator` (whatever that ends up being named) instead of being handled here - the only part of this that's relevant to authentication is whether the user is a collaborator or not, not the specific status code. This is part of the more detailed refactor I suggested above.

File: src/backends/github/AuthenticationPage.js [hunk 14]
```diff
@@ -16,7 +18,7 @@ export default class AuthenticationPage extends React.Component {
     const cfg = {
       base_url: this.props.base_url,
       site_id: (document.location.host.split(':')[0] === 'localhost') ? 'cms.netlify.com' : this.props.siteId
-    }
+    };
```
Comment (Benaiah) [inline src/backends/github/AuthenticationPage.js]: This change, while good, is unrelated to the PR.

File: src/backends/github/API.js (hunk 10 above)
Comment (erquhart) [inline src/backends/github/API.js]: Great breakdown @Benaiah, I'm on board with this.

File: .gitignore (hunk 8 above)
Comment (erquhart) [inline .gitignore]: I personally think it's fine. `manifest.yml` is a Bluemix artifact, and `.imdone` is from Atom. A lengthy `.gitignore` isn't a bad thing if it improves the developer experience for contributors.

@tortilaman what is `config.local.yml` from?

Comment (tortilaman) [inline .gitignore]: `config.local.yml` is storing my backend config so I don't have to go searching for it to test with GitHub.

File: src/backends/github/API.js (hunk 10 above)
Comment (tortilaman) [inline src/backends/github/API.js]: @Benaiah Would `isCollaborator` then need to follow promise syntax and utilize `resolve()` etc.? I'm of the understanding that without doing that `.then((response)...` will still work based on what I have now.

File: src/backends/github/AuthenticationPage.js (hunk 5 above)
Comment (tortilaman) [inline src/backends/github/AuthenticationPage.js]: Totally a mistake. Was going to define another proptype, but then removed it and forgot to remove the comma.

File: .gitignore (hunk 8 above)
Comment (erquhart) [inline .gitignore]: @tortilaman let's leave that value out since it's not a standard file, but the other two can stay.

**Comment by erquhart** [comment]
@Benaiah leaving this to you, merge once you're happy with the changes.

**Comment by erquhart** [comment]
@tortilaman merge conflict.

File: .gitignore (hunk 8 above)
Comment (Benaiah) [inline .gitignore]: @erquhart sounds like a solid policy to me, thanks for the clarification.

**Comment by erquhart** [comment]
@Benaiah also, we should squash-merge.

**Comment by tortilaman** [comment]
Apologies, I was definitely going to squash this on my last commit, but the interactive rebase got quite complicated with the merge and figured I would have to re-order commits and such, and I just figured it would be better left to you.

**Comment by erquhart** [comment]
No problem! Squash merging is a simple solution, no sweat.

**Comment by tortilaman** [comment]
Just wanted to check if you'd like any other changes, or if this should be merged @erquhart @Benaiah.

**Comment by erquhart** [comment]
@tortilaman sounds good, reach out if we can help.

**Comment by tortilaman** [comment]
Alright @erquhart @Benaiah I think this is ready for a squash merge.

File: src/backends/github/API.js [hunk 15]
```diff
@@ -67,6 +80,9 @@ export default class API {
       if (contentType && contentType.match(/json/)) {
         return this.parseJsonResponse(response);
       }
+      if (responseStatus !== 200) {
```
Comment (Benaiah) [inline src/backends/github/API.js]: A generic request method like this should probably be checking the equivalent of `response.ok`, which is "status in the range 200-299" ([response.ok on MDN](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok), see also [MDN's list of 2xx status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#Successful_responses)). Changing it to this line would fix that:

```js
      if ((responseStatus >= 200) && (responseStatus <= 299)) {
```

Comment (tortilaman) [inline src/backends/github/API.js]: That's leftover from the collaborator check. I switched to a write permission check, so I'll just remove that actually.

**Comment by tortilaman** [comment]
Alright, everything resolved.

**Comment by Benaiah** [comment]
@tortilaman thanks so much!

**Comment by tortilaman** [comment]
Thanks for all the feedback!
